test/jdk/tools/sincechecker/SinceChecker.java

```diff
@@ -0,0 +1,769 @@
+        }
+    }
+
+    private void analyzeClassRecord(TypeElement te, String version, Types types, Elements elements) {
+        Set<Modifier> classModifiers = te.getModifiers();
+        if (!(classModifiers.contains(Modifier.PUBLIC) || classModifiers.contains(Modifier.PROTECTED))) {
+            return;
+        }
+        persistElement(te.getEnclosingElement(), te, types, version);
+        elements.getAllMembers(te).stream()
+                .filter(element -> element.getModifiers().contains(Modifier.PUBLIC) || element.getModifiers().contains(Modifier.PROTECTED))
+                .filter(element -> element.getKind().isField()
+                        || element.getKind() == ElementKind.METHOD
+                        || element.getKind() == ElementKind.CONSTRUCTOR)
+                .forEach(element -> persistElement(te, element, types, version));
+        te.getEnclosedElements().stream()
+                .filter(element -> element.getKind().isDeclaredType())
+                .map(TypeElement.class::cast)
+                .forEach(nestedClass -> analyzeClassRecord(nestedClass, version, types, elements));
+    }
+
+    public void persistElement(Element explicitOwner, Element element, Types types, String version) {
+        String uniqueId = getElementName(explicitOwner, element, types);
+        IntroducedIn introduced = classDictionary.computeIfAbsent(uniqueId, i -> new IntroducedIn());
+        if (isPreview(element, uniqueId, version)) {
+            if (introduced.introducedPreview == null) {
+                introduced.introducedPreview = version;
+            }
+        } else {
+            if (introduced.introducedStable == null) {
+                introduced.introducedStable = version;
+            }
+        }
+    }
+
+    private boolean isPreview(Element el, String uniqueId, String currentVersion) {
+        while (el != null) {
+            Symbol s = (Symbol) el;
+            if ((s.flags() & Flags.PREVIEW_API) != 0) {
+                return true;
+            }
+            el = el.getEnclosingElement();
+        }
+
+        return LEGACY_PREVIEW_METHODS.containsKey(currentVersion)
+                && LEGACY_PREVIEW_METHODS.get(currentVersion).contains(uniqueId);
+    }
+
+    private void testThisModule(String moduleName) throws Exception {
+        Path home = Paths.get(System.getProperty("java.home"));
+        Path srcZip = home.resolve("lib").resolve("src.zip");
+        if (Files.notExists(srcZip)) {
+            //possibly running over an exploded JDK build, attempt to find a
+            //co-located full JDK image with src.zip:
+            Path testJdk = Paths.get(System.getProperty("test.jdk"));
+            srcZip = testJdk.getParent().resolve("images").resolve("jdk").resolve("lib").resolve("src.zip");
+        }
+        File f = new File(srcZip.toUri());
+        if (!f.exists() && !f.isDirectory()) {
+            throw new SkippedException("Skipping Test because src.zip wasn't found");
+        }
```

Comment:
For modern/new code I would recommend using `java.nio.file.Path` instead of `java.io.File`.

I see you actually have a `Path` in `srcZip`, so use `Files.exists(srcZip)` and `Files.isDirectory(srcZip)`

Reply:
Fixed in 32edb4d811978f87bd2f850c214191e8fc7b8c88